quark/subproject.py

```diff
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
+                self.gitlab_token_header = "PRIVATE-TOKEN"
+                self.gitlab_token_is_pat = True  # Is this a Personal/Private Access Token? (PAT)
+                return
+
+        # If we're running in a CI job, use the CI_JOB_TOKEN.
+        if "CI_JOB_TOKEN" in os.environ:
+            self.gitlab_token = os.environ["CI_JOB_TOKEN"]
+            self.gitlab_token_header = "JOB-TOKEN"
+            self.gitlab_token_is_pat = False  # Is this a Personal/Private Access Token? (PAT)
+            return
+
+        raise QuarkError("Missing authentication token.")
+
+    def _parse_url(self, url):
+        fragments = Subproject._parse_fragment(url) if url.fragment else {}
+
+        # There are mainly used for logging messages.
+        self.parsed_job = None
+        self.parsed_ref = None
+        self.parsed_pkg = None
+
+        self.stamp["sha1"] = fragments.get("sha1")
+
+        if url.scheme == "gitlab+ci":
+            if "job" not in fragments:
+                raise QuarkError("Missing job fragment in URL: " + url)
+
+            if "ref" in fragments and "job" not in fragments:
+                raise QuarkError("Missing job fragment in URL when ref is given: " + url)
+
+            parts = url.path.split("/")
+
+            self.parsed_project_name = "/".join(parts[0 : parts.index("artifacts")]).strip("/")
+            self.parsed_artifact_name = parts[-1]
+            self.parsed_job = fragments.get("job")
+            self.parsed_ref = fragments.get("ref")
+
+            artifact_path = "/".join(parts[parts.index("artifacts") + 1 :])
+
+            if "ref" in fragments and not self.gitlab_token_is_pat:
+                # NOTE: We have "ref" and "job" but we don't have a Personal Access Token (PAT). We
+                # can't resolve the job ID using the API and a CI job token can't be used for this.
+                #
+                # TODO: Find a way to resolve the job ID without requiring private tokens so that we
+                # can use a single endpoint (and we can always store the job ID inside the stamp
+                # file).
+                self.parsed_endpoint_url = "/projects/%s/jobs/artifacts/%s/raw/%s?job=%s" % (
+                    urllib.parse.quote_plus(self.parsed_project_name),
+                    urllib.parse.quote_plus(fragments["ref"]),
+                    artifact_path,
+                    urllib.parse.quote_plus(fragments["job"]),
+                )
+            else:
+                if "ref" in fragments and self.gitlab_token_is_pat:
+                    print_msg("resolving job id for " + self.parsed_artifact_name, self._comment())
+                    self.stamp["job_id"] = self._resolve_job_id(
+                        self.parsed_project_name,
+                        fragments["ref"],
+                        fragments["job"],
+                    )
+                else:
+                    self.stamp["job_id"] = str(fragments["job"])
```

Comment:
So ultimately `job_id` can either be the job name (if we don't have a PAT), or the actual job ID? Couldn't we at least distinguish the two cases explicitly (e.g. leaving `job_id` at `None` when we don't know it)? Otherwise the rest of the code has no indication about this fact, and is left to guess.